### PR TITLE
[Module] Cloudflare

### PR DESCRIPTION
**Pull request by samkahchiin**
### Summary
This module done the following:

- Domain validation
  - Configure a CNAME record in DNS configuration to establish control of your domain name. This allows ACM to automatically renew DNS-validated certificates before they expire, as long as the DNS record has not changed
- Route to defaults routes
    - api.my-project.com to the NLB
    - bastion.my-project.com
    - cdn.my-project.com to the S3-Bucket (using Cloudflare Workers)
    - xx.my-project.com to the S3-Bucket (using Cloudflare Workers) for any subdomain needed

### CODE REVIEW

File: cloudflare/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+## Pre-setup
+1. Setup and publish your CDN cloudflare worker.
+- Some useful references:
+  - Quickstart: https://developers.cloudflare.com/workers/get-started/quickstarts/
+  - Cloudflare CDN tutorials: https://developers.cloudflare.com/workers/tutorials/configure-your-cdn/
+  - Introduction to Cloudflare Worker: https://egghead.io/courses/introduction-to-cloudflare-workers-5aa3
+  - Cloudflare Router: https://github.com/dbl-works/cloudflare-router
+
+2. Make sure you have create API token in your cloudflare account with sufficient permissions
```
Comment (swiknaba) [inline cloudflare/README.md:21]:
```suggestion
2. Make sure you have created an API token in your Cloudflare account with sufficient permissions
```

File: cloudflare/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+```terraform
+module "cloudflare" {
+  source = "github.com/dbl-works/terraform//cloudflare?ref=v2022.05.26"
+
+  domain = "example.com"
+  subject_alternative_names = ["*.example.com"]
+  bastion_public_dns = "project-staging-xxxxx.elb.eu-central-1.amazonaws.com"
+  nlb_dns_name = "project-staging-xxxxxxx.eu-central-1.elb.amazonaws.com"
+  cdn_worker_script_name = "serve-cdn"
+  s3_cdn_buckets = [
```
Comment (swiknaba) [inline cloudflare/README.md:37]:
```suggestion
  s3_public_buckets = [
```

we've renamed the "cdn" bucket module into "public" to be more general. should we do the same here?

File: cloudflare/main.tf [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+resource "cloudflare_record" "api" {
+  zone_id = cloudflare_zone.default.id
+  name    = "api"
+  type    = "CNAME"
+  value   = var.nlb_dns_name
+  proxied = true
+}
+
+# bastion.my-project.com
+resource "cloudflare_record" "bastion" {
```
Comment (swiknaba) [inline cloudflare/main.tf:44]: should we make this optional, e.g. by skipping this resource, if `bastion_public_dns` is not given?

**Review by swiknaba — CHANGES_REQUESTED** [review]
pls have a look 🙂 

https://github.com/dbl-works/terraform/pull/87
https://github.com/dbl-works/terraform/runs/6867926154?check_suite_focus=true

image

**Comment by samkahchiin** [comment]
@swiknaba 
Have fixed accordingly. 
Just realised that the domain_validation_options is not available in data source (Only available in resources).  😢

**Pull request closed without merging**